components/vision.py

```diff
@@ -1,36 +1,115 @@
+import time
+
 from networktables import NetworkTables
-from typing import Tuple
+from typing import Optional
+from dataclasses import dataclass
+from magicbot import feedback
+
+
+@dataclass
+class VisionData:
+    #: The distance to the target in metres.
+    distance: float
+
+    #: The angle to the target in radians.
+    angle: float
+
+    #: An arbitrary timestamp, in seconds,
+    #: for when the vision system last obtained data.
+    timestamp: float
+
+    __slots__ = ("distance", "angle", "timestamp")
 
 
 class Vision:
+    @property
+    def ping_time(self) -> float:
+        return self.ping_time_entry.getDouble(0.0)
+
+    @ping_time.setter
+    def ping_time(self, value: float) -> None:
+        self.ping_time_entry.setDouble(value)
+
+    @property
+    def rio_pong_time(self) -> float:
+        return self.rio_pong_time_entry.getDouble(0.0)
+
+    @property
+    def raspi_pong_time(self) -> float:
+        return self.raspi_pong_time_entry.getDouble(0.0)
+
+    @property
+    def latency(self) -> float:
+        return self.latency_entry.getDouble(0.0)
+
+    @latency.setter
+    def latency(self, value: float) -> None:
+        self.latency_entry.setDouble(value)
+
+    @property
+    def processing_time(self) -> float:
+        return self.processing_time_entry.getDouble(0.0)
+
+    @processing_time.setter
+    def processing_time(self, value: float) -> None:
+        self.processing_time_entry.setDouble(value)
+
     def __init__(self) -> None:
+        self.last_pong = time.monotonic()
```

Comment:
You would construct the delegate comms class here, and it would do this and most of the following setup.